{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload sequencing data to SRA\n",
    "This Python Jupyter notebook uploads the sequencing data to the NIH [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra), or SRA."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create BioProject and BioSamples\n",
    "The first step was done manually to create the BioProject and BioSamples.\n",
    "Note that for new future uploads related to the RBD DMS, you may be able to use the existing BioProject, but since this is the first entries in these project I needed to create a new BioProject.\n",
    "\n",
    "To create these, I went to the [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra) and signed in using the box at the upper right of the webpage, and then went to the [SRA Submission Portal](https://submit.ncbi.nlm.nih.gov/subs/sra/).\n",
    "I then manually completed the first five steps, which define the project and samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create submission sheet\n",
    "The sixth step is to create the submission sheet in `*.tsv` format, which is done by the following code.\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ftplib\n",
    "import os\n",
    "import tarfile\n",
    "import datetime\n",
    "\n",
    "import natsort\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the PacBio runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading PacBio runs from ../data/PacBio_runs.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>run</th>\n",
       "      <th>subreads</th>\n",
       "      <th>ccs_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>200415_A</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib1</td>\n",
       "      <td>200415_B</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib2</td>\n",
       "      <td>200415_A</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib2</td>\n",
       "      <td>200415_B</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library       run                                           subreads  \\\n",
       "0    lib1  200415_A  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "1    lib1  200415_B  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "2    lib2  200415_A  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "3    lib2  200415_B  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "\n",
       "                                    ccs_file  \n",
       "0  ../results/ccs/lib1_200415_A_ccs.fastq.gz  \n",
       "1  ../results/ccs/lib1_200415_B_ccs.fastq.gz  \n",
       "2  ../results/ccs/lib2_200415_A_ccs.fastq.gz  \n",
       "3  ../results/ccs/lib2_200415_B_ccs.fastq.gz  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio_runs_file = os.path.join('../', config['pacbio_runs'])\n",
    "\n",
    "print(f\"Reading PacBio runs from {pacbio_runs_file}\")\n",
    "\n",
    "pacbio_runs = (\n",
    "    pd.read_csv(pacbio_runs_file)\n",
    "    .assign(ccs_file=lambda x: f\"../{config['ccs_dir']}/\" + x['library'] + '_' + x['run'] + '_ccs.fastq.gz')\n",
    "    )\n",
    "\n",
    "pacbio_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the PacBio CCSs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name        library_ID  \\\n",
       "0  PacBio_CCSs  lib1_PacBio_CCSs   \n",
       "1  PacBio_CCSs  lib1_PacBio_CCSs   \n",
       "2  PacBio_CCSs  lib2_PacBio_CCSs   \n",
       "3  PacBio_CCSs  lib2_PacBio_CCSs   \n",
       "\n",
       "                                               title     library_strategy  \\\n",
       "0  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "1  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "2  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "3  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "\n",
       "  library_source   library_selection library_layout     platform  \\\n",
       "0      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "1      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "2      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "3      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "\n",
       "  instrument_model                                 design_description  \\\n",
       "0    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "1    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "2    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "3    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "\n",
       "  filetype                          filename_fullpath  \n",
       "0    fastq  ../results/ccs/lib1_200415_A_ccs.fastq.gz  \n",
       "1    fastq  ../results/ccs/lib1_200415_B_ccs.fastq.gz  \n",
       "2    fastq  ../results/ccs/lib2_200415_A_ccs.fastq.gz  \n",
       "3    fastq  ../results/ccs/lib2_200415_B_ccs.fastq.gz  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio_submissions = (\n",
    "    pacbio_runs\n",
    "    .assign(\n",
    "        sample_name='PacBio_CCSs',  # BioSample created in SRA wizard\n",
    "        library_ID=lambda x: x['library'] + '_PacBio_CCSs',  # unique library ID\n",
    "        title='PacBio CCSs linking variants to barcodes for SARS-CoV-2 RBD deep mutational scanning',\n",
    "        library_strategy='Synthetic-Long-Read',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='Restriction Digest',\n",
    "        library_layout='single',\n",
    "        platform='PACBIO_SMRT',\n",
    "        instrument_model='PacBio Sequel',\n",
    "        design_description='Restriction digest of plasmids carrying barcoded RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['ccs_file'],      \n",
    "        )\n",
    "    .drop(columns=pacbio_runs.columns)\n",
    "    )\n",
    "\n",
    "pacbio_submissions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from ../data/barcode_runs.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sort_bin</th>\n",
       "      <th>concentration</th>\n",
       "      <th>date</th>\n",
       "      <th>number_cells</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>6600000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin2</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>3060000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin3</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>2511000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin4</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>2992000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib2</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>6420000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0953...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library        sample sample_type  sort_bin  concentration    date  \\\n",
       "0    lib1  SortSeq_bin1     SortSeq         1            NaN  200416   \n",
       "1    lib1  SortSeq_bin2     SortSeq         2            NaN  200416   \n",
       "2    lib1  SortSeq_bin3     SortSeq         3            NaN  200416   \n",
       "3    lib1  SortSeq_bin4     SortSeq         4            NaN  200416   \n",
       "4    lib2  SortSeq_bin1     SortSeq         1            NaN  200416   \n",
       "\n",
       "   number_cells                                                 R1  \n",
       "0       6600000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "1       3060000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "2       2511000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "3       2992000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "4       6420000  /shared/ngs/illumina/tstarr/200427_D00300_0953...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "illumina_runs_file = os.path.join('../', config['barcode_runs'])\n",
    "\n",
    "print(f\"Reading Illumina runs from {illumina_runs_file}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(illumina_runs_file)\n",
    "\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for Illumina data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name         library_ID  \\\n",
       "0  expression_barcodes  lib1_SortSeq_bin1   \n",
       "1  expression_barcodes  lib1_SortSeq_bin1   \n",
       "2  expression_barcodes  lib1_SortSeq_bin1   \n",
       "3  expression_barcodes  lib1_SortSeq_bin1   \n",
       "4  expression_barcodes  lib1_SortSeq_bin1   \n",
       "\n",
       "                                               title library_strategy  \\\n",
       "0  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "1  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "2  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "3  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "4  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "\n",
       "  library_source library_selection library_layout  platform  \\\n",
       "0      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description filetype  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                   filename_fullpath  \n",
       "0  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "1  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "2  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "3  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "4  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "illumina_submissions = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        sample_name=lambda x: x['sample_type'].map({'SortSeq': 'expression_barcodes',\n",
    "                                                    'TiteSeq': 'hACE2_binding_barcodes'}),\n",
    "        library_ID=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'SARS-CoV-2 RBD deep mutational scanning Illumina barcode sequencing for ' + x['sample'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='Illumina HiSeq 2500',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';'),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip())\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "illumina_submissions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now concatenate the PacBio and Illumina submissions into tidy format (one line per file), make sure all the files exist, and also make short name versions of them that lack the path:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>file_exists</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "      <td>True</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name         library_ID  \\\n",
       "0          PacBio_CCSs   lib1_PacBio_CCSs   \n",
       "1          PacBio_CCSs   lib1_PacBio_CCSs   \n",
       "2          PacBio_CCSs   lib2_PacBio_CCSs   \n",
       "3          PacBio_CCSs   lib2_PacBio_CCSs   \n",
       "4  expression_barcodes  lib1_SortSeq_bin1   \n",
       "\n",
       "                                               title     library_strategy  \\\n",
       "0  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "1  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "2  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "3  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "4  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "\n",
       "  library_source   library_selection library_layout     platform  \\\n",
       "0      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "1      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "2      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "3      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "4      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "\n",
       "      instrument_model                                 design_description  \\\n",
       "0        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "1        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "2        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "3        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "4  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "\n",
       "  filetype                                  filename_fullpath  file_exists  \\\n",
       "0    fastq          ../results/ccs/lib1_200415_A_ccs.fastq.gz         True   \n",
       "1    fastq          ../results/ccs/lib1_200415_B_ccs.fastq.gz         True   \n",
       "2    fastq          ../results/ccs/lib2_200415_A_ccs.fastq.gz         True   \n",
       "3    fastq          ../results/ccs/lib2_200415_B_ccs.fastq.gz         True   \n",
       "4    fastq  /shared/ngs/illumina/tstarr/200427_D00300_0952...         True   \n",
       "\n",
       "                                            filename  \n",
       "0                         lib1_200415_A_ccs.fastq.gz  \n",
       "1                         lib1_200415_B_ccs.fastq.gz  \n",
       "2                         lib2_200415_A_ccs.fastq.gz  \n",
       "3                         lib2_200415_B_ccs.fastq.gz  \n",
       "4  200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    pd.concat([pacbio_submissions, illumina_submissions], ignore_index=True)\n",
    "    .assign(file_exists=lambda x: x['filename_fullpath'].map(os.path.isfile),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename),\n",
    "            )\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['file_exists'].all(), submissions_tidy.query('file_exists == False')\n",
    "\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the actual submission, we need a \"wide\" data frame that for each unique `sample_name` / `library_ID` gives all of the files each in different columns.\n",
    "These should be files without the full path.\n",
    "\n",
    "First, look at how many files there are for each sample / library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>n_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib2_SortSeq_bin1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib1_TiteSeq_12_bin1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_13_bin1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib1_TiteSeq_10_bin4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib1_TiteSeq_10_bin3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib1_TiteSeq_10_bin2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_04_bin3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_name            library_ID  n_files\n",
       "0       expression_barcodes     lib1_SortSeq_bin1       10\n",
       "1       expression_barcodes     lib2_SortSeq_bin1        8\n",
       "2    hACE2_binding_barcodes  lib1_TiteSeq_12_bin1        6\n",
       "3    hACE2_binding_barcodes  lib2_TiteSeq_16_bin1        6\n",
       "4    hACE2_binding_barcodes  lib2_TiteSeq_13_bin1        6\n",
       "..                      ...                   ...      ...\n",
       "133  hACE2_binding_barcodes  lib1_TiteSeq_10_bin4        2\n",
       "134  hACE2_binding_barcodes  lib1_TiteSeq_10_bin3        2\n",
       "135  hACE2_binding_barcodes  lib1_TiteSeq_10_bin2        2\n",
       "136  hACE2_binding_barcodes  lib2_TiteSeq_16_bin4        2\n",
       "137  hACE2_binding_barcodes  lib2_TiteSeq_04_bin3        1\n",
       "\n",
       "[138 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .groupby(['sample_name', 'library_ID'])\n",
    " .aggregate(n_files=pd.NamedAgg('filename_fullpath', 'count'))\n",
    " .sort_values('n_files', ascending=False)\n",
    " .reset_index()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the wide submission data frame.\n",
    "Note we keep only the filename column with the path lacking the full directory information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>...</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>filename3</th>\n",
       "      <th>filename4</th>\n",
       "      <th>filename5</th>\n",
       "      <th>filename6</th>\n",
       "      <th>filename7</th>\n",
       "      <th>filename8</th>\n",
       "      <th>filename9</th>\n",
       "      <th>filename10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>...</td>\n",
       "      <td>lib1_200415_A_ccs.fastq.gz</td>\n",
       "      <td>lib1_200415_B_ccs.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>...</td>\n",
       "      <td>lib2_200415_A_ccs.fastq.gz</td>\n",
       "      <td>lib2_200415_B_ccs.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_003.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_004.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_005.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_002.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_003.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_004.fast...</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_005.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin2</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200416_lib1_FITCbin2_TGGCTTCA_L001_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin2_TGGCTTCA_L001_R1_002.fast...</td>\n",
       "      <td>200416_lib1_FITCbin2_TGGCTTCA_L002_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin2_TGGCTTCA_L002_R1_002.fast...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin3</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200416_lib1_FITCbin3_TGGTGGTA_L001_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin3_TGGTGGTA_L001_R1_002.fast...</td>\n",
       "      <td>200416_lib1_FITCbin3_TGGTGGTA_L002_R1_001.fast...</td>\n",
       "      <td>200416_lib1_FITCbin3_TGGTGGTA_L002_R1_002.fast...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_15_bin4</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200422_s15-b4_TAGGATGA_L001_R1_001.fastq.gz</td>\n",
       "      <td>200422_s15-b4_TAGGATGA_L002_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L001_R1_001.fastq.gz</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L001_R1_002.fastq.gz</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L001_R1_003.fastq.gz</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L002_R1_001.fastq.gz</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L002_R1_002.fastq.gz</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L002_R1_003.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin2</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200422_s16-b2_TCCGTCTA_L001_R1_001.fastq.gz</td>\n",
       "      <td>200422_s16-b2_TCCGTCTA_L002_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin3</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200422_s16-b3_TCTTCACA_L001_R1_001.fastq.gz</td>\n",
       "      <td>200422_s16-b3_TCTTCACA_L002_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin4</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>200422_s16-b4_TGAAGAGA_L001_R1_001.fastq.gz</td>\n",
       "      <td>200422_s16-b4_TGAAGAGA_L002_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_name            library_ID  \\\n",
       "0               PacBio_CCSs      lib1_PacBio_CCSs   \n",
       "1               PacBio_CCSs      lib2_PacBio_CCSs   \n",
       "2       expression_barcodes     lib1_SortSeq_bin1   \n",
       "3       expression_barcodes     lib1_SortSeq_bin2   \n",
       "4       expression_barcodes     lib1_SortSeq_bin3   \n",
       "..                      ...                   ...   \n",
       "133  hACE2_binding_barcodes  lib2_TiteSeq_15_bin4   \n",
       "134  hACE2_binding_barcodes  lib2_TiteSeq_16_bin1   \n",
       "135  hACE2_binding_barcodes  lib2_TiteSeq_16_bin2   \n",
       "136  hACE2_binding_barcodes  lib2_TiteSeq_16_bin3   \n",
       "137  hACE2_binding_barcodes  lib2_TiteSeq_16_bin4   \n",
       "\n",
       "                                                 title     library_strategy  \\\n",
       "0    PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "1    PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "2    SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "3    SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "4    SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "..                                                 ...                  ...   \n",
       "133  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "134  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "135  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "136  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "137  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "\n",
       "    library_source   library_selection library_layout     platform  \\\n",
       "0        SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "1        SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "2        SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "3        SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "4        SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "..             ...                 ...            ...          ...   \n",
       "133      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "134      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "135      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "136      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "137      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "\n",
       "        instrument_model                                 design_description  \\\n",
       "0          PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "1          PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "2    Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "3    Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "4    Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "..                   ...                                                ...   \n",
       "133  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "134  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "135  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "136  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "137  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "\n",
       "     ...                                           filename  \\\n",
       "0    ...                         lib1_200415_A_ccs.fastq.gz   \n",
       "1    ...                         lib2_200415_A_ccs.fastq.gz   \n",
       "2    ...  200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...   \n",
       "3    ...  200416_lib1_FITCbin2_TGGCTTCA_L001_R1_001.fast...   \n",
       "4    ...  200416_lib1_FITCbin3_TGGTGGTA_L001_R1_001.fast...   \n",
       "..   ...                                                ...   \n",
       "133  ...        200422_s15-b4_TAGGATGA_L001_R1_001.fastq.gz   \n",
       "134  ...        200422_s16-b1_TATCAGCA_L001_R1_001.fastq.gz   \n",
       "135  ...        200422_s16-b2_TCCGTCTA_L001_R1_001.fastq.gz   \n",
       "136  ...        200422_s16-b3_TCTTCACA_L001_R1_001.fastq.gz   \n",
       "137  ...        200422_s16-b4_TGAAGAGA_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                             filename2  \\\n",
       "0                           lib1_200415_B_ccs.fastq.gz   \n",
       "1                           lib2_200415_B_ccs.fastq.gz   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fast...   \n",
       "3    200416_lib1_FITCbin2_TGGCTTCA_L001_R1_002.fast...   \n",
       "4    200416_lib1_FITCbin3_TGGTGGTA_L001_R1_002.fast...   \n",
       "..                                                 ...   \n",
       "133        200422_s15-b4_TAGGATGA_L002_R1_001.fastq.gz   \n",
       "134        200422_s16-b1_TATCAGCA_L001_R1_002.fastq.gz   \n",
       "135        200422_s16-b2_TCCGTCTA_L002_R1_001.fastq.gz   \n",
       "136        200422_s16-b3_TCTTCACA_L002_R1_001.fastq.gz   \n",
       "137        200422_s16-b4_TGAAGAGA_L002_R1_001.fastq.gz   \n",
       "\n",
       "                                             filename3  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L001_R1_003.fast...   \n",
       "3    200416_lib1_FITCbin2_TGGCTTCA_L002_R1_001.fast...   \n",
       "4    200416_lib1_FITCbin3_TGGTGGTA_L002_R1_001.fast...   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134        200422_s16-b1_TATCAGCA_L001_R1_003.fastq.gz   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename4  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L001_R1_004.fast...   \n",
       "3    200416_lib1_FITCbin2_TGGCTTCA_L002_R1_002.fast...   \n",
       "4    200416_lib1_FITCbin3_TGGTGGTA_L002_R1_002.fast...   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134        200422_s16-b1_TATCAGCA_L002_R1_001.fastq.gz   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename5  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L001_R1_005.fast...   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134        200422_s16-b1_TATCAGCA_L002_R1_002.fastq.gz   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename6  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L002_R1_001.fast...   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134        200422_s16-b1_TATCAGCA_L002_R1_003.fastq.gz   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename7  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L002_R1_002.fast...   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename8  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L002_R1_003.fast...   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                             filename9  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L002_R1_004.fast...   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "133                                                NaN   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "\n",
       "                                            filename10  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2    200416_lib1_FITCbin1_TGGAACAA_L002_R1_005.fast...  \n",
       "3                                                  NaN  \n",
       "4                                                  NaN  \n",
       "..                                                 ...  \n",
       "133                                                NaN  \n",
       "134                                                NaN  \n",
       "135                                                NaN  \n",
       "136                                                NaN  \n",
       "137                                                NaN  \n",
       "\n",
       "[138 rows x 21 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .assign(\n",
    "        filename_count=lambda x: x.groupby(['sample_name', 'library_ID'])['filename'].cumcount() + 1,\n",
    "        filename_col=lambda x: 'filename' + x['filename_count'].map(lambda c: str(c) if c > 1 else '')\n",
    "        )\n",
    "    .pivot(\n",
    "        index='library_ID',\n",
    "        columns='filename_col',\n",
    "        values='filename',\n",
    "        )\n",
    "    )\n",
    "\n",
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .drop(columns=['filename_fullpath', 'file_exists', 'filename'])\n",
    "    .drop_duplicates()\n",
    "    .merge(submissions_wide[natsort.natsorted(submissions_wide.columns)],\n",
    "           on='library_ID',\n",
    "           validate='one_to_one',\n",
    "           )\n",
    "    )\n",
    "\n",
    "submissions_wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now write the wide submissions data frame to a `*.tsv` file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "submissions_spreadsheet = 'SRA_submission_spreadsheet.tsv'\n",
    "\n",
    "submissions_wide.to_csv(submissions_spreadsheet, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This submission sheet was then manually uploaded in Step 6 of the SRA submission wizard (*SRA metadata*)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload the actual files\n",
    "Step 7 of the SRA submission wizard is to upload the files.\n",
    "In order to do this, we first make a `*.tar` file with all of the files.\n",
    "Since this takes a long time, we only create the file if it doesn't already exist, so it is only created the first time this notebook is run.\n",
    "**Note that this will cause a problem if you add more sequencing files to upload after running the notebook, in that case the cell below will need to altered.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA_submission.tar already exists, not creating it again\n"
     ]
    }
   ],
   "source": [
    "tar_filename = 'SRA_submission.tar'\n",
    "\n",
    "if os.path.isfile(tar_filename):\n",
    "    print(f\"{tar_filename} already exists, not creating it again\")\n",
    "else:\n",
    "    try:\n",
    "        with tarfile.open(tar_filename, mode='w') as f:\n",
    "            for i, tup in enumerate(submissions_tidy.itertuples()):\n",
    "                print(f\"Adding file {i + 1} of {len(submissions_tidy)} to {tar_filename}\")\n",
    "                f.add(tup.filename_fullpath, arcname=tup.filename)\n",
    "            print(f\"Added all files to {tar_filename}\")\n",
    "    except:\n",
    "        if os.path.isfile(tar_filename):\n",
    "            os.remove(tar_filename)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See the size of the `*.tar` file to upload and make sure it has the expected files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The size of SRA_submission.tar is 26.2 GB\n",
      "SRA_submission.tar contains all 376 expected files.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The size of {tar_filename} is {os.path.getsize(tar_filename) / 1e9:.1f} GB\")\n",
    "\n",
    "with tarfile.open(tar_filename) as f:\n",
    "    files_in_tar = set(f.getnames())\n",
    "if files_in_tar == set(submissions_tidy['filename']):\n",
    "    print(f\"{tar_filename} contains all {len(files_in_tar)} expected files.\")\n",
    "else:\n",
    "    raise ValueError(f\"{tar_filename} does not have all the expected files.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The SRA instructions then give several ways to upload; we will do it using the FTP method.\n",
    "First, specify the FTP address, username, password, and subfolder given by the SRA submission wizard instructions.\n",
    "In order to avoid having the password be public here, that is in a separate text file that is **not** included in the GitHub repo (so this needs to be run in Jesse's directory that has this password):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the following are provided by SRA wizard insturctions\n",
    "ftp_address = 'ftp-private.ncbi.nlm.nih.gov'\n",
    "ftp_username = 'subftp'\n",
    "ftp_account_folder = 'uploads/jbloom_fhcrc.org_IuMBgK44'\n",
    "with open('ftp_password.txt') as f:\n",
    "    ftp_password = f.read().strip()\n",
    "    \n",
    "# meaningful name for subfolder\n",
    "ftp_subfolder = 'SARS-CoV-2-RBD_DMS'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now create FTP connection and upload the TAR file.\n",
    "Note that this takes a while.\n",
    "If you are worried that it will timeout given the size of your file, you can run this notebook via `slurm` so there is no timing out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting upload at 2020-06-16 18:07:42.589806\n",
      "Finished upload at 2020-06-17 00:45:05.078738\n"
     ]
    }
   ],
   "source": [
    "print(f\"Starting upload at {datetime.datetime.now()}\")\n",
    "\n",
    "with ftplib.FTP(ftp_address) as ftp:\n",
    "    ftp.login(user=ftp_username,\n",
    "              passwd=ftp_password,\n",
    "              )\n",
    "    ftp.cwd(ftp_account_folder)\n",
    "    ftp.mkd(ftp_subfolder)\n",
    "    ftp.cwd(ftp_subfolder)\n",
    "    with open(tar_filename, 'rb') as f:\n",
    "        ftp.storbinary(f\"STOR {tar_filename}\", f)\n",
    "        \n",
    "print(f\"Finished upload at {datetime.datetime.now()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, used the SRA wizard to select the `*.tar` archive and complete the submission.\n",
    "Note that there is a warning of missing files since everything was uploaded as a `*.tar` rather than individual files.\n",
    "They should all be found when you hit the button to proceed and the `*.tar` is unpacked.\n",
    "\n",
    "There was then a message that the submission was processing, and data would be released immediately upon processing.\n",
    "The submission number is `SUB7594564`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
